qiskit/circuit/parameterexpression.py

```diff
@@ -232,6 +232,40 @@ def _apply_operation(self, operation: Callable,
 
         return ParameterExpression(parameter_symbols, expr)
 
+    def gradient(self, param: 'ParameterExpression') -> Union['ParameterExpression', float]:
```

Comment:
For consistency with the rest of `ParameterExpresison`, it would be better if this always returned a `ParameterExpression`, even if there are no remaining free symbols. (This way, downstream operations won't have to check if the object is a float or a `ParameterExpression`.)

Reply:
Thansk @kdk for your comments.
We did try to always return a 'ParameterExpression' at first but experienced the case/issue where a 'ParameterExpression' did not have any more symbols which it could bind, e.g. for the gradient of ´4x´.